### PR TITLE
Update matter-lock driver for U-tec

### PR DESCRIPTION
**Pull request by HunsupJung**
Check all that apply

# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [ ] I have reviewed the README.md file
          - [ ] I have reviewed the CODE_OF_CONDUCT.md file
          - [ ] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [ ] New feature
- [ ] Refactor

# Checklist

- [ ] I have performed a self-review of my code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have verified my changes by testing with a device or have communicated a plan for testing
- [ ] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change
https://smartthings.atlassian.net/browse/IOTE-4705

# Summary of Completed Tests

### CODE REVIEW

**Comment by nickolas-deboom** [comment]
After an initial review, the changes look good to me overall. I have two comments - (1) are there additional profiles needed for the possible combinations of lock/user/pin/schedule + battery/batteryLevel? Or are we just providing profiles for known configurations at this time? (2) Would you be able to create a few test cases similar to what I have here? https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/pull/1796/files#diff-188bbf519a245b9967a0a86552cf7b20bef0a2b5ce6598196cca7e6591acc56d

In order to show that the `BatPercentRemaining` attribute being present results in the correct profile being chosen.

**Comment by HunsupJung** [comment]
>(1) are there additional profiles needed for the possible combinations of lock/user/pin/schedule + battery/batteryLevel? Or are we just providing profiles for known configurations at this time?

There may be other possible combinations. But I added lock-user-pin-batteryLevel, lock-user-pin-battery and lock-user-pin-scheduled-battery which are highly likely. For U-tec, the lock-user-pin-batteryLevel profile is enough. 
Do you want to consider all edge cases?

>(2) Would you be able to create a few test cases similar to what I have here?

Okay, I will update it.

**Comment by nickolas-deboom** [comment]
> > (1) are there additional profiles needed for the possible combinations of lock/user/pin/schedule + battery/batteryLevel? Or are we just providing profiles for known configurations at this time?
> 
> There may be other possible combinations. But I added lock-user-pin-batteryLevel, lock-user-pin-battery and lock-user-pin-scheduled-battery which are highly likely. For U-tec, the lock-user-pin-batteryLevel profile is enough. Do you want to consider all edge cases?
> 
> > (2) Would you be able to create a few test cases similar to what I have here?
> 
> Okay, I will update it.

I think it's ok to include just the likely ones for now rather than every combination.

File: drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua [hunk 1]
```diff
@@ -1685,7 +1752,10 @@ local new_matter_lock_handler = {
     capabilities.lock,
     capabilities.lockUsers,
     capabilities.lockCredentials,
-    capabilities.lockSchedules
+    capabilities.lockSchedules,
+    capabilities.remoteControlStatus,
```
Comment (ctowns) [inline drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua:1756]: There is no handler for `remoteControlStatus`, should there be or should this be removed?

File: drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  subscribe_request:merge(DoorLock.attributes.RequirePINforRemoteOperation:subscribe(mock_device))
+  subscribe_request:merge(DoorLock.events.LockOperation:subscribe(mock_device))
+  subscribe_request:merge(DoorLock.events.DoorLockAlarm:subscribe(mock_device))
+  subscribe_request:merge(DoorLock.events.LockUserChange:subscribe(mock_device))
+  test.socket["matter"]:__expect_send({mock_device.id, subscribe_request})
+  test.mock_device.add_test_device(mock_device)
+end
+test.set_test_init_function(test_init)
+
+test.register_coroutine_test(
```
Comment (ctowns) [inline drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua:77]: Could you create a test case for the following cases also?
1) battery percent remaining is available ✅ 
2) battery charge level only is available ❌ 
3) battery charge level and battery percent remaining are available ❌ 

You will just need to modify the attributes list in the `AttributeList:build_test_report_data` function you are using to test these different combinations of attributes in the new test cases.

File: drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua [hunk 3]
```diff
@@ -340,6 +359,49 @@ local function max_year_schedule_of_user_handler(driver, device, ib, response)
...
 end
 
+---------------------------------
+-- Power Source Attribute List --
+---------------------------------
+local function handle_power_source_attribute_list(driver, device, ib, response)
+  local support_battery_percentage = false
+  for _, attr in ipairs(ib.data.elements) do
+    -- Re-profile the device if BatPercentRemaining (Attribute ID 0x0C) is present.
+    if attr.value == 0x0C then
```
Comment (ctowns) [inline drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua:369]: Can we add an explicit check for the `BatChargeLevel` attribute here?
```
if attr.value == 0x0C then
  support_battery_percentage = true
end
if attr.value == 0x0E then
    support_battery_level = true
end
------then later add this :-----------
  if support_battery_percentage then
    profile_name = profile_name .. "-battery"
  elseif support_battery_level
    profile_name = profile_name .. "-batteryLevel"
  end

**Review by ctowns — COMMENTED** [review]
This looks pretty good to me! After adding the additional unit tests and making the small change in the handler, I think this will be good to go - thank you @HunsupJung !

Also, has this been regression tested with the other locks that use this driver?

**Comment by Kwang-Hui** [comment]
I tested this PR with Aqara U200/U300 door lock - it was fine. Thank you.

File: drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua (hunk 1 above)
Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua:1756]: I don't think it's needed. I will remove it.

File: drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua (hunk 3 above)
Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua:369]: Okay, I will

File: drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua (hunk 2 above)
Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua:77]: I didn't understand your explanation.
There are first cases and second one below. And there's no different between first one and third one.
I do not know which test case to upload additionally.

Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua:77]: I thought the second case was added, but it wasn't. I will modify it.

Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/src/test/test_new_matter_lock_battery.lua:77]: For reference, it is impossible to support only `Battery percent remaining`. Because if server supports BAT feature, `Battery change level` is mandatory. I will create as follows.

1. Attributes related to BAT feature is not available.
2. battery charge level only is available.
3. battery charge level and battery percent remaining are available.

**Comment by Kwang-Hui** [comment]
@ctowns could you review this pr again?

File: drivers/SmartThings/matter-lock/profiles/lock-user-pin-schedule-battery.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+name: lock-user-pin-schedule-battery
```
Comment (hcarter-775) [inline drivers/SmartThings/matter-lock/profiles/lock-user-pin-schedule-battery.yml:1]: for completion, could we also add this profile with -batteryLevel?

Comment (HunsupJung) [reply drivers/SmartThings/matter-lock/profiles/lock-user-pin-schedule-battery.yml:1]: Sure. I will add the lock-user-pin-schedule-batteryLevel profile.

**Comment by Kwang-Hui** [comment]
@hcarter-775 Could you please review again?

**Review by hcarter-775 — APPROVED** [review]
looks good! just rebase and resolve conflicts :+1:
